README.md
<div align="center">
  <img width="225" src="https://github.com/piotrmurach/strings/blob/master/assets/strings_logo.png" alt="strings logo" />
</div>

# Strings::Numeral

[![Gem Version](https://badge.fury.io/rb/strings-numeral.svg)][gem]
[![Actions CI](https://github.com/piotrmurach/strings-numeral/workflows/CI/badge.svg?branch=master)][gh_actions_ci]
[![Build status](https://ci.appveyor.com/api/projects/status/494htkcankqegwtg?svg=true)][appveyor]
[![Maintainability](https://api.codeclimate.com/v1/badges/de0c5ad1cba6715b7135/maintainability)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/github/piotrmurach/strings-numeral/badge.svg?branch=master)][coverage]
[![Inline docs](https://inch-ci.org/github/piotrmurach/strings-numeral.svg?branch=master)][inchpages]

[gem]: https://badge.fury.io/rb/strings-numeral
[gh_actions_ci]: https://github.com/piotrmurach/strings-numeral/actions?query=workflow%3ACI
[appveyor]: https://ci.appveyor.com/project/piotrmurach/strings-numeral
[codeclimate]: https://codeclimate.com/github/piotrmurach/strings-numeral/maintainability
[coverage]: https://coveralls.io/github/piotrmurach/strings-numeral?branch=master
[inchpages]: https://inch-ci.org/github/piotrmurach/strings-numeral

> Express numbers as string numerals.

**Strings::Numeral** provides conversions of numbers to numerals component for [Strings](https://github.com/piotrmurach/strings).

## Features

* No monkey-patching String class
* Functional API that can be easily wrapped by other objects
* Instance based configuration
* Highly performant

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'strings-numeral'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install strings-numeral

## Contents

* [1. Usage](#1-usage)
* [2. API](#2-api)
  * [2.1 numeralize](#21-numeralize)
  * [2.2 cardinalize](#22-cardinalize)
  * [2.3 ordinalize](#23-ordinalize)
  * [2.4 monetize](#24-monetize)
  * [2.5 romanize](#25-romanize)
  * [2.6 configuration](#26-configuration)
* [3. Extending core classes](#3-extending-core-classes)

## 1. Usage

**Strings::Numeral** helps to express any number as a numeral in words. It exposes few methods to achieve this. For example, you can express a number as a cardinal numeral using `cardinalize`:

```ruby
Strings::Numeral.cardinalize(1234)
# => "one thousand, two hundred thirty four"
```

But you're not limited to converting integers only. It can handle decimals as well:

```ruby
Strings::Numeral.cardinalize(1234.567)
# => "one thousand, two hundred thirty four and five hundred sixty seven thousandths"
```

For more options on how to customize formatting see [configuration](#25-configuration) section.

Similarly, you can convert a number to a ordinal numeral with `ordinalize`:

```ruby
Strings::Numeral.ordinalize(1234)
# => "one thousand, two hundred thirty fourth"
```

You can also convert a number to a short ordinal:

```ruby
Strings::Numeral.ordinalize(1234, short: true)
# => "1234th"
```

Using `monetize` you can convert any number into a monetary numeral:

```ruby
Strings::Numeral.monetize(1234.567)
# => "one thousand, two hundred thirty four dollars and fifty seven cents",
```

To turn a number into a roman numeral use `romanize`:

```ruby
Strings::Numeral.romanize(2020)
# => "MMXX"
```

## 2. API

### 2.1 numeralize

The `normalize` is a wrapping method for the [cardinalize](#22-cardinalize) and [ordinalize](#23-ordinalize) methods. By default it converts a number to cardinal numeral:

```ruby
Strings::Numeral.numeralize(1234.567)
# => "one thousand, two hundred thirty four and five hundred sixty seven thousandths"
```

You can also make it convert to ordinal numerals using `:term` option:

```ruby
Strings::Numeral.numeralize(1234.567, term: :ord)
# => "one thousand, two hundred thirty fourth and five hundred sixty seven thousandths"
```

### 2.2 cardinalize

To express a number as a cardinal numeral use `cardinalize` or `cardinalise`.

```ruby
Strings::Numeral.cardinalize(1234)
# => "one thousand, two hundred thirty four"
```

You're not limited to integers only. You can also express decimal numbers as well:

```ruby
Strings::Numeral.cardinalize(123.456)
# => "one hundred twenty three and four hundred fifty six thousandths"
```

By default the fractional part of a decimal number is expressed as a fraction. If you wish to spell out fractional part digit by digit use `:decimal` option with `:digit` value:

```ruby
Strings::Numeral.cardinalize(123.456, decimal: :digit)
# => "one hundred twenty three point four five six"
```

You may prefer to use a different delimiter for thousand's. You can do use by passing the `:delimiter` option:

```ruby
Strings::Numeral.cardinalize(1_234_567, delimiter: " and ")
# => "one million and two hundred thirty four thousand and five hundred sixty seven"
```

To change word that splits integer from factional part use `:separator` option:

```ruby
Strings::Numeral.cardinalize(1_234.567, separator: "dot")
# => "one thousand, two hundred thirty four dot five hundred sixty seven thousandths"
```

### 2.3 ordinalize

To express a number as a cardinal numeral use `ordinalize` or `ordinalise`.

```ruby
Strings::Numeral.ordinalize(1234)
# => "one thousand, two hundred thirty fourth"
```

You're not limited to integers only. You can also express decimal numbers as well:

```ruby
Strings::Numeral.ordinalize(123.456)
# => "one hundred twenty third and four hundred fifty six thousandths"
```

By default the fractional part of a decimal number is expressed as a fraction. If you wish to spell out fractional part digit by digit use `:decimal` option with `:digit` value:

```ruby
Strings::Numeral.ordinalize(123.456, decimal: :digit)
# => "one hundred twenty third point four five six"
```

You may prefer to use a different delimiter for thousand's. You can do use by passing the `:delimiter` option:

```ruby
Strings::Numeral.ordinalize(1_234_567, delimiter: " and ")
# => "one million and two hundred thirty four thousand and five hundred sixty seventh"
```

To change word that splits integer from factional part use `:separator` option:

```ruby
Strings::Numeral.ordinalize(1_234.567, separator: "dot")
# => "one thousand, two hundred thirty fourth dot five hundred sixty seven thousandths"
```

### 2.4 monetize

To express a number as a monetary numeral use `monetize` or `monetise`.

```ruby
Strings::Numeral.monetize(123.456)
# => "one hundred twenty three dollars and forty six cents",
```

By default `monetize` displays money using `USD` currency. You can change this with the `:currency` option that as value accepts internationally recognised symbols. Currently support currencies are: `EUR`, `GBP`, `JPY`, `PLN` and `USD`.

```ruby
Strings::Numeral.monetize(123.456, currency: :jpy)
# => "one hundred twenty three yen and forty six sen"
```

### 2.5 romanize

To convert a number into a Roman numeral use `romanize`:

```ruby
Strings::Numeral.romanize(2020)
# => "MMXX"
```

### 2.6 configuration

All available configuration settings are:

* `currency` - Adds currency words for integer and fractional parts. Supports `EUR`, `GBP`, `JPY`, `PLN` and `USD`. Defaults to `USD`.
* `decimal` - Formats fractional part of a number. The `:digit` value spells out every digit and the `:fraction` appends divider word. Defaults to `:fraction`.
* `delimiter` - Sets the thousands delimiter. Defaults to `", "`.
* `separator` - Sets the separator between the fractional and integer parts. Defaults to `"and"` for `:fraction` and `"point"` for `:digit` option.
* `strict` - Enables number validation for the input parameter. Defaults to `false`.
* `trailing_zeros` - If `true` keeps trailing zeros at the end of the fractional part. Defaults to `false`.

The above settings can be passed as keyword arguments:

```ruby
Strings::Numeral.cardinalize("12.100", trailing_zeros: true, decimal: :digit)
# => "twelve point one zero zero"
```

Or you can configure the settings for an instance during initialisation:

```ruby
numeral = Strings::Numeral.new(delimiter: "; ", separator: "dot")
```

After initialisation, you can use `configure` to change settings inside a block:

```ruby
numeral.configure do |config|
  config.delimiter "; "
  config.separator "dot"
  config.decimal :digit
  config.trailing_zeros true
end
```

Once configured, you can use the instance like so:

```ruby
numeral.cardinalize("1234.56700")
# => "one thousand; two hundred thirty four dot five six seven zero zero"
```

## 3. Extending Core Classes

Though it is highly discouraged to pollute core Ruby classes, you can add the required methods to `String`, `Float` and `Integer` classes using refinements.

For example, if you wish to only extend `Float` class with `cardinalize` method do:

```ruby
module MyFloatExt
  refine Float do
    def cardinalize(**options)
      Strings::Numeral.cardinalize(self, **options)
    end
  end
end
```

Then `cardinalize` method will be available for any float number where refinement is applied:

```ruby
using MyFloatExt

12.34.cardinalize
# => "twelve and thirty four"
```

However, if you want to include all the **Strings::Numeral** methods in `Float`, `Integer` and `String` classes, you can use provided extensions file:


```ruby
require "strings/numeral/extensions"

using Strings::Numeral::Extensions
```

Alternatively, you can choose what class you wish to refine with all the methods:

```ruby
require "bigdecimal"
require "strings/numeral/extensions"

module MyBigDecimalExt
  refine BigDecimal do
    include Strings::Numeral::Extensions::Methods
  end
end

using MyBigDecimalExt
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/piotrmurach/strings-numeral. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/piotrmurach/strings-numeral/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Strings::Numeral project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/piotrmurach/strings-numeral/blob/master/CODE_OF_CONDUCT.md).

## Copyright

Copyright (c) 2019 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
# Change log

## [v0.2.0] - unreleased

### Added
* Add the ability to configure settings after initialisation
* Add strict mode setting for enabling number validation of input parameter

### Changed
* Change to use forwardable to delegate methods
* Change extensions to require core library when loaded
* Change conversion methods to accept only numbers in strict mode

## [v0.1.0] - 2019-12-18

* Initial implementation and release

[v0.2.0]: https://github.com/piotrmurach/strings-numeral/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/piotrmurach/strings-numeral/compare/5546406...v0.1.0


spec/spec_helper.rb
# frozen_string_literal: true

if ENV["COVERAGE"] == "true"
  require "simplecov"
  require "coveralls"

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ])

  SimpleCov.start do
    command_name "spec"
    add_filter "spec"
  end
end

require "bundler/setup"
require "strings/numeral"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/unit/config_spec.rb
# frozen_string_literal: true

RSpec.describe Strings::Numeral, "configuration" do
  it "defaults configuration values" do
    numeral = Strings::Numeral.new

    expect(numeral.configuration.currency).to eq(:usd)
    expect(numeral.configuration.decimal).to eq(:fraction)
    expect(numeral.configuration.delimiter).to eq(", ")
    expect(numeral.configuration.separator).to eq(nil)
    expect(numeral.configuration.strict).to eq(false)
    expect(numeral.configuration.trailing_zeros).to eq(false)
  end

  it "configures settings at initialisation" do
    numeral = Strings::Numeral.new(
      currency: "PLN",
      decimal: :digit,
      delimiter: "; ",
      separator: "dot",
      strict: true,
      trailing_zeros: true
    )

    expect(numeral.configuration.currency).to eq("PLN")
    expect(numeral.configuration.decimal).to eq(:digit)
    expect(numeral.configuration.delimiter).to eq("; ")
    expect(numeral.configuration.separator).to eq("dot")
    expect(numeral.configuration.strict).to eq(true)
    expect(numeral.configuration.trailing_zeros).to eq(true)

    expect(numeral.monetize("1234.56700"))
      .to eq("one thousand; two hundred thirty four zlotys dot five seven groszy")
  end

  it "configures strict mode and validates value as not a number" do
    numeral = Strings::Numeral.new(strict: true)

    expect {
      numeral.monetize("strict")
    }.to raise_error(Strings::Numeral::Error, "not a number: \"strict\"")
  end

  it "configures settings at runtime using keyword arguments" do
    numeral = Strings::Numeral.new

    numeral.configure(currency: "PLN", decimal: :digit, delimiter: "; ",
                      separator: "dot", strict: true, trailing_zeros: true)

    expect(numeral.configuration.currency).to eq("PLN")
    expect(numeral.configuration.decimal).to eq(:digit)
    expect(numeral.configuration.delimiter).to eq("; ")
    expect(numeral.configuration.separator).to eq("dot")
    expect(numeral.configuration.strict).to eq(true)
    expect(numeral.configuration.trailing_zeros).to eq(true)

    expect(numeral.monetize("1234.56700"))
      .to eq("one thousand; two hundred thirty four zlotys dot five seven groszy")
  end

  it "configures settings at runtime for a cardinal instance" do
    numeral = Strings::Numeral.new

    numeral.configure do |config|
      config.decimal :digit
      config.delimiter "; "
      config.separator "dot"
      config.strict true
      config.trailing_zeros true
    end

    expect(numeral.configuration.decimal).to eq(:digit)
    expect(numeral.configuration.delimiter).to eq("; ")
    expect(numeral.configuration.separator).to eq("dot")
    expect(numeral.configuration.strict).to eq(true)
    expect(numeral.configuration.trailing_zeros).to eq(true)

    expect(numeral.cardinalize("1234.56700"))
      .to eq("one thousand; two hundred thirty four dot five six seven zero zero")
  end

  it "configures settings at runtime for an ordinal instance" do
    numeral = Strings::Numeral.new

    numeral.configure do |config|
      config.decimal :digit
      config.delimiter "; "
      config.separator "dot"
      config.trailing_zeros true
    end

    expect(numeral.configuration.decimal).to eq(:digit)
    expect(numeral.configuration.delimiter).to eq("; ")
    expect(numeral.configuration.separator).to eq("dot")
    expect(numeral.configuration.trailing_zeros).to eq(true)

    expect(numeral.ordinalize("1234.56700"))
      .to eq("one thousand; two hundred thirty fourth dot five six seven zero zero")
  end

  it "configures settings at runtime for monetization" do
    numeral = Strings::Numeral.new

    numeral.configure do |config|
      config.currency :pln
      config.decimal :digit
      config.delimiter "; "
      config.separator "dot"
    end

    expect(numeral.configuration.currency).to eq(:pln)
    expect(numeral.configuration.decimal).to eq(:digit)
    expect(numeral.configuration.delimiter).to eq("; ")
    expect(numeral.configuration.separator).to eq("dot")

    expect(numeral.monetize("1234.56700"))
      .to eq("one thousand; two hundred thirty four zlotys dot five seven groszy")
  end
end


spec/unit/monetize_spec.rb
# frozen_string_literal: true

RSpec.describe Strings::Numeral, "#monetize" do
  {
    1 => "one dollar",
    20 => "twenty dollars",
    21 => "twenty one dollars",
    100 => "one hundred dollars",
    1234 => "one thousand, two hundred thirty four dollars",
    12345 => "twelve thousand, three hundred forty five dollars",
    123456 => "one hundred twenty three thousand, four hundred fifty six dollars",
    1234567 => "one million, two hundred thirty four thousand, five hundred sixty seven dollars",
    -125 => "negative one hundred twenty five dollars",
    0.1 => "zero dollars and ten cents",
    0.7 => "zero dollars and seventy cents",
    0.01 => "zero dollars and one cent",
    0.21 => "zero dollars and twenty one cents",
    0.1234 => "zero dollars and twelve cents",
    12.100 => "twelve dollars and ten cents",
    12.000 => "twelve dollars",
    123.456 => "one hundred twenty three dollars and forty six cents",
    -114.5678 => "negative one hundred fourteen dollars and fifty seven cents",
    1234.567 => "one thousand, two hundred thirty four dollars and fifty seven cents",
    -3456.07 => "negative three thousand, four hundred fifty six dollars and seven cents"
  }.each do |num, word|
    it "monetizes #{num.inspect} to #{word.inspect}" do
      expect(Strings::Numeral.monetize(num)).to eq(word)
    end
  end

  it "monetizes 123.45 in EUR " do
    expect(Strings::Numeral.monetize(123.456, currency: :eur))
      .to eq("one hundred twenty three euros and forty six cents")
  end

  it "monetizes 123.45 in GBP " do
    expect(Strings::Numeral.monetize(123.456, currency: :gbp))
      .to eq("one hundred twenty three pounds and forty six pence")
  end

  it "monetizes 123.45 in JPY " do
    expect(Strings::Numeral.monetize(123.456, currency: :jpy))
      .to eq("one hundred twenty three yen and forty six sen")
  end

  it "monetizes 123.45 in PLN " do
    expect(Strings::Numeral.monetize(123.456, currency: :pln))
      .to eq("one hundred twenty three zlotys and forty six groszy")
  end

  it "checks value in strict mode as not a number" do
    expect {
      Strings::Numeral.monetize("unknown", strict: true)
    }.to raise_error(Strings::Numeral::Error, "not a number: \"unknown\"")
  end
end


lib/strings/numeral/extensions.rb
# frozen_string_literal: true

require_relative "../numeral" unless defined?(Strings::Numeral)

module Strings
  class Numeral
    module Extensions
      Methods = Module.new do
        def numeralize(**options)
          Strings::Numeral.numeralize(self, **options)
        end

        def cardinalize(**options)
          Strings::Numeral.cardinalize(self, **options)
        end

        def ordinalize(**options)
          Strings::Numeral.ordinalize(self, **options)
        end

        def ordinalize_short
          Strings::Numeral.ordinalize_short(self)
        end

        def monetize(**options)
          Strings::Numeral.monetize(self, **options)
        end

        def romanize
          Strings::Numeral.romanize(self)
        end
      end

      refine String do
        include Methods
      end

      refine Float do
        include Methods
      end

      refine Integer do
        include Methods
      end
    end # Extensions
  end # Numeral
end # Strings


lib/strings/numeral.rb
# frozen_string_literal: true

require "forwardable"

require_relative "numeral/configuration"
require_relative "numeral/version"

module Strings
  class Numeral
    class Error < StandardError; end

    NEGATIVE = "negative"
    HUNDRED = "hundred"
    ZERO = "zero"
    AND = "and"
    POINT = "point"
    SPACE = " "

    CARDINALS = {
      0 => "",
      1 => "one",
      2 => "two",
      3 => "three",
      4 => "four",
      5 => "five",
      6 => "six",
      7 => "seven",
      8 => "eight",
      9 => "nine",
      10 => "ten",
      11 => "eleven",
      12 => "twelve",
      13 => "thirteen",
      14 => "fourteen",
      15 => "fifteen",
      16 => "sixteen",
      17 => "seventeen",
      18 => "eighteen",
      19 => "nineteen",
      20 => "twenty",
      30 => "thirty",
      40 => "forty",
      50 => "fifty",
      60 => "sixty",
      70 => "seventy",
      80 => "eighty",
      90 => "ninety"
    }.freeze

    CARDINAL_TO_SHORT_ORDINAL = {
      0 => "th",
      1 => "st",
      11 => "th",
      2 => "nd",
      12 => "th",
      3 => "rd",
      13 => "th",
      4 => "th",
      5 => "th",
      6 => "th",
      7 => "th",
      8 => "th",
      9 => "th"
    }.freeze

    CARDINAL_TO_ORDINAL = {
      "zero" => "zeroth",
      "one" => "first",
      "two" => "second",
      "three" => "third",
      "four" => "fourth",
      "five" => "fifth",
      "six" => "sixth",
      "seven" => "seventh",
      "eight" => "eighth",
      "nine" => "ninth",
      "ten" => "tenth",
      "eleven" => "eleventh",
      "twelve" => "twelfth",
      "thirteen" => "thirteenth",
      "fourteen" => "fourteenth",
      "fifteen" => "fifteenth",
      "sixteen" => "sixteenth",
      "seventeen" => "seventeenth",
      "eighteen" => "eighteenth",
      "nineteen" => "nineteenth",
      "twenty" => "twentieth",
      "thirty" => "thirtieth",
      "forty" => "fortieth",
      "fifty" => "fiftieth",
      "sixty" => "sixtieth",
      "seventy" => "seventieth",
      "eighty" => "eightieth",
      "ninety" => "ninetieth"
    }.freeze

    CARDINAL_TO_ROMAN = {
      1 => "I",
      4 => "IV",
      5 => "V",
      9 => "IX",
      10 => "X",
      40 => "XL",
      50 => "L",
      90 => "XC",
      100 => "C",
      400 => "CD",
      500 => "D",
      900 => "CM",
      1000 => "M"
    }.freeze

    SCALES = [
      "hundreds-tens-ones",
      "thousand",
      "million",
      "billion",
      "trillion",
      "quadrillion",
      "quintillion",
      "sextillion",
      "septillion",
      "octillion",
      "nonillion",
      "decillion",
      "undecillion",
      "duodecillion",
      "tredecillion",
      "quattuordecillion",
      "quindecillion",
      "sexdecillion",
      "septemdecillion",
      "octodecillion",
      "novemdecillion",
      "vigintillion"
    ].freeze

    DECIMAL_SLOTS = [
      "tenths",
      "hundredths",
      "thousandths",
      "ten-thousandths",
      "hundred-thousandths",
      "millionths",
      "ten-millionths",
      "hundred-millionths",
      "billionths",
      "ten-billionths",
      "hundred-billionths",
      "trillionths",
      "quadrillionths",
      "quintillionths",
      "sextillionths",
      "septillionths",
      "octillionths",
      "nonillionths",
      "decillionths",
      "undecillionths",
      "duodecillionths",
      "tredecillionths",
      "quattuordecillionths",
      "quindecillionths",
      "sexdecillionths",
      "septemdecillionths",
      "octodecillionths",
      "novemdecillionths",
      "vigintillionths"
    ].freeze

    CURRENCIES = {
      eur: {
        unit: "euro",
        units: "euros",
        decimal_unit: "cent",
        decimal_units: "cents"
      },
      gbp: {
        unit: "pound",
        units: "pounds",
        decimal_unit: "pence",
        decimal_units: "pence"
      },
      jpy: {
        unit: "yen",
        units: "yen",
        decimal_unit: "sen",
        decimal_units: "sen"
      },
      pln: {
        unit: "zloty",
        units: "zlotys",
        decimal_unit: "grosz",
        decimal_units: "groszy"
      },
      usd: {
        unit: "dollar",
        units: "dollars",
        decimal_unit: "cent",
        decimal_units: "cents"
      }
    }.freeze

    # Global instance
    #
    # @api private
    def self.instance
      @instance ||= Numeral.new
    end

    class << self
      extend Forwardable

      delegate %i[numeralize cardinalize cardinalise ordinalize
                  ordinalise ordinalize_short monetize monetise
                  romanize romanise] => :instance
    end

    # Create numeral with custom configuration
    #
    # @return [Numeral]
    #
    # @api public
    def initialize(**options)
      configuration.update(**options)
    end

    # Access configuration
    #
    # @api public
    def configuration
      @configuration ||= Configuration.new
    end

    # Configure numerals settings
    #
    # @example
    #   numeral = Strings::Numeral.new
    #   numeral.configure do |config|
    #     config.decimal :digit
    #     config.delimiter "; "
    #     config.separator "dot"
    #     config.trailing_zeros true
    #   end
    #
    # @example
    #   numeral = Strings::Numeral.new
    #   numeral.configure decimal: :digit, separator: "dot"
    #
    # @yieldparam [Configuration]
    #
    # @api public
    def configure(**options)
      if block_given?
        yield configuration
      else
        configuration.update(**options)
      end
    end

    # Convert a number to a numeral
    #
    # @param [Numeric,String] num
    #   the number to convert
    #
    # @api public
    def numeralize(num, **options)
      case options.delete(:term)
      when /ord/
        ordinalize(num, **options)
      else
        cardinalize(num, **options)
      end
    end

    # Convert a number to a cardinal numeral
    #
    # @example
    #   cardinalize(1234)
    #   # => one thousand, two hundred thirty four
    #
    # @param [Numeric,String] num
    #
    # @return [String]
    #
    # @api public
    def cardinalize(num, **options)
      check_number(num, **options)
      convert_numeral(num, **options)
    end
    alias cardinalise cardinalize

    # Convert a number to an ordinal numeral
    #
    # @example
    #   ordinalize(1234)
    #   # => one thousand, two hundred thirty fourth
    #
    #   ordinalize(12, short: true) # => 12th
    #
    # @param [Numeric,String] num
    #   the number to convert
    #
    # @return [String]
    #
    # @api public
    def ordinalize(num, **options)
      check_number(num, **options)
      if options[:short]
        ordinalize_short(num)
      else
        decimals = (num.to_i.abs != num.to_f.abs)
        sentence = convert_numeral(num, **options)
        separators = [AND, POINT,
                      options.fetch(:separator, @configuration.separator)].compact

        if decimals && sentence =~ /(\w+) (#{Regexp.union(separators)})/
          last_digits = $1
          separator = $2
          replacement = CARDINAL_TO_ORDINAL[last_digits]
          pattern = /#{last_digits} #{separator}/
          suffix = "#{replacement} #{separator}"
        elsif sentence =~ /(\w+)$/
          last_digits = $1
          replacement = CARDINAL_TO_ORDINAL[last_digits]
          pattern = /#{last_digits}$/
          suffix = replacement
        end

        if replacement
          sentence.sub(pattern, suffix)
        else
          sentence
        end
      end
    end
    alias ordinalise ordinalize

    # Convert a number to a short ordinal form
    #
    # @example
    #   ordinalize_short(123) # => 123rd
    #
    # @param [Numeric, String] num
    #   the number to convert
    #
    # @return [String]
    #
    # @api private
    def ordinalize_short(num)
      num_abs = num.to_i.abs

      num.to_i.to_s + (CARDINAL_TO_SHORT_ORDINAL[num_abs % 100] ||
        CARDINAL_TO_SHORT_ORDINAL[num_abs % 10])
    end

    # Convert a number into a monetary numeral
    #
    # @example
    #   monetize(123.45)
    #   # => "one hundred twenty three dollars and forty five cents"
    #
    # @param [Numeric,String] num
    #   the number to convert
    #
    # @return [String]
    #
    # @api public
    def monetize(num, **options)
      check_number(num, **options)
      sep = options.fetch(:separator, @configuration.separator)
      curr_name = options.fetch(:currency, @configuration.currency)
      n = format("%0.2f", num.to_s)
      decimals = (num.to_i.abs != num.to_f.abs)
      sentence = convert_numeral(n, **options.merge(trailing_zeros: true))
      dec_num = n.split(".")[1]
      curr = CURRENCIES[curr_name.to_s.downcase.to_sym]
      separators = [AND, POINT, sep].compact

      if decimals
        regex = /(\w+) (#{Regexp.union(separators)})/
        sentence.sub!(regex, "\\1 #{curr[:units]} \\2")
      else
        sentence += SPACE + (num.to_i == 1 ? curr[:unit] : curr[:units])
      end

      if decimals
        slots = Regexp.union(DECIMAL_SLOTS.map { |slot| slot.chomp("s") })
        regex = /(#{slots})s?/i
        suffix = dec_num.to_i == 1 ? curr[:decimal_unit] : curr[:decimal_units]
        if sentence.sub!(regex, suffix).nil?
          sentence += SPACE + suffix
        end
      end

      sentence
    end
    alias monetise monetize

    # Convert a number to a roman numeral
    #
    # @example
    #   romanize(2020) # => "MMXX"
    #
    # @param [Integer] num
    #   the number to convert
    # @param [Boolean] strict
    #   whether or not to validate input is a number
    #
    # @return [String]
    #
    # @api public
    def romanize(num, strict: configuration.strict)
      check_number(num, strict: strict)
      n = num.to_i

      if n < 1 || n > 4999
        raise Error, "'#{n}' is out of range"
      end

      CARDINAL_TO_ROMAN.keys.reverse_each.reduce([]) do |word, card|
        while n >= card
          n -= card
          word << CARDINAL_TO_ROMAN[card]
        end
        word
      end.join
    end

    private

    # Check whether the value is a number when in strict mode
    #
    # @api private
    def check_number(value, **options)
      strict = options.fetch(:strict, @configuration.strict)
      strict && !number?(value) && raise_not_number(value)
    end

    # Check whether or not value is a number
    #
    # @param [Object] value
    #   the value to check
    #
    # @return [Boolean]
    #
    # @api private
    def number?(value)
      !Float(value).nil?
    rescue TypeError, ArgumentError
      false
    end

    # Raise not a number error
    #
    # @raise [Error]
    #
    # @api private
    def raise_not_number(value)
      raise Error, "not a number: #{value.inspect}"
    end

    # Convert a number into a numeral
    #
    # @param [Numeric] num
    #   the number to convert to numeral
    # @param [String] delimiter
    #   sets the thousand's delimiter, defaults to `, `
    # @param [String] decimal
    #   the decimal word conversion, defaults to `:fraction`
    # @param [String] separator
    #   sets the separator between the fractional and integer numerals,
    #   defaults to `and` for fractions and `point` for digits
    #
    # @return [String]
    #   the number as numeral
    #
    # @api private
    def convert_numeral(num, **options)
      delimiter = options.fetch(:delimiter, @configuration.delimiter)
      decimal = options.fetch(:decimal, @configuration.decimal)
      separator = options.fetch(:separator, @configuration.separator)

      negative = num.to_i < 0
      n = num.to_i.abs
      decimals = (n != num.to_f.abs)

      sentence = convert_to_words(n).join(delimiter)

      if sentence.empty?
        sentence = ZERO
      end

      if negative
        sentence = NEGATIVE + SPACE + sentence
      end

      if decimals
        sep = separator.nil? ? (decimal == :fraction ? AND : POINT) : separator
        sentence += SPACE + sep + SPACE + convert_decimals(num, **options)
      end

      sentence
    end

    # Convert decimal part to words
    #
    # @param [String] trailing_zeros
    #   whether or not to keep trailing zeros, defaults to `false`
    #
    # @return [String]
    #
    # @api private
    def convert_decimals(num, **options)
      delimiter = options.fetch(:delimiter, @configuration.delimiter)
      decimal = options.fetch(:decimal, @configuration.decimal)
      trailing_zeros = options.fetch(:trailing_zeros, @configuration.trailing_zeros)

      dec_num = num.to_s.split(".")[1]
      dec_num.gsub!(/0+$/, "") unless trailing_zeros

      case decimal
      when :fraction
        unit = DECIMAL_SLOTS[dec_num.to_s.length - 1]
        unit = unit[0...-1] if dec_num.to_i == 1 # strip off 's'
        convert_to_words(dec_num.to_i).join(delimiter) + SPACE + unit
      when :digit
        dec_num.chars.map do |n|
          (v = convert_tens(n.to_i)).empty? ? ZERO : v
        end.join(SPACE)
      else
        raise Error, "Unknown decimal option '#{decimal.inspect}'"
      end
    end

    # Convert an integer to number words
    #
    # @param [Integer] n
    #
    # @return [Array[String]]
    #
    # @api public
    def convert_to_words(n)
      words = []

      SCALES.each_with_index do |scale, i|
        mod = n % 1000

        word = []
        word << convert_hundreds(mod)
        word << scale unless i.zero?

        words.insert(0, word.join(SPACE))

        n /= 1000

        break if n.zero?
      end

      words
    end

    # Convert 3 digit number to equivalent word
    #
    # @return [String]
    #
    # @api private
    def convert_hundreds(num)
      word = []
      hundreds = (num % 1000) / 100
      tens = num % 100

      if !hundreds.zero?
        word << convert_tens(hundreds)
        word << HUNDRED
      end

      if !tens.zero?
        word << convert_tens(tens)
      end

      word.join(SPACE)
    end

    # Convert number in 0..99 range to equivalent word
    #
    # @return [String]
    #
    # @api private
    def convert_tens(num)
      word = []
      tens = num % 100

      if tens.to_s.size < 2 || tens <= 20
        word << CARDINALS[tens]
      else
        word << CARDINALS[(tens / 10) * 10]
        word << CARDINALS[tens % 10] unless (tens % 10).zero?
      end

      word.join(SPACE)
    end
  end # Numeral
end # Strings
